Add various terrain type subtables

babaissarkar opened the pull request:
Resolves #9969.
These expose `terrain_type::mvt_type()` and `terrain_type::def_type()` as `mvt_alias` and `def_alias` subtables (array) to `wesnoth.terrain_types` table, which have the same meaning as mentioned in [TerrainWML](https://wiki.wesnoth.org/TerrainWML).

(Note: no way to figure out whether it is a `best of` or `worst of` for the latter two yet, because it's a bit too involved, but may be done in future.)

Vultraz commented on src/help/help_topic_generators.cpp:
```diff
@@ -163,14 +158,45 @@ static std::string print_behavior_description(
 	return ss.str();
 }
 
+static std::vector<std::string> get_special_notes(const terrain_type& type) {
```
Should be in an anon namespace instead of `static`.

babaissarkar replied:
Moved all `static` functions to anon namespace.

soliton- replied:
What is this based on? Seems like a pointless change.

babaissarkar replied:
> What is this based on? Seems like a pointless change.

What is pointless, the moving of `static` functions to anon namespace, or me separating out the part that assigned special notes into a function?

soliton- replied:
The requested change here from static to anon namespace.

Vultraz replied:
We generally put free-standing functions and variables in an anon namespace instead of declaring them static... I recommended the change for consistency.

soliton- replied:
We generally do that based on hundreds of existing static declarations? Is anyone else aware of this? If this is meant seriously then I suggest to document this somewhere with reasoning so it's not just a random back and forth.

babaissarkar replied:
This is a suitable place for adding this info: https://wiki.wesnoth.org/CodingStandard

babaissarkar commented:
Example lua console output:
```lua
wesnoth.terrain_mask wesnoth.terrain_types
$ wesnoth.terrain_types['Sm']
{id='quagmire',def_terrains={'swamp_water'},editor_image='terrain/swamp/mud-tile.png',light=0,icon='',mvt_terrains={'swamp_water'},keep=false,healing=0,basic_terrains={'swamp_water'},editor_name=Muddy Quagmire,village=false,castle=false,description=Swamp,name=Swamp}
$ wesnoth.terrain_types['Sm^Bw/']
{id='quagmire^bridgediag1',def_terrains={'swamp_water','flat'},editor_image='terrain/swamp/mud-tile.png~BLIT(terrain/bridge/wood-ne-sw.png)',light=0,icon='',mvt_terrains={'swamp_water','flat'},keep=false,healing=0,basic_terrains={'flat','swamp_water'},editor_name=Muddy Quagmire / Wooden Bridge,village=false,castle=false,description=Bridge,name=Bridge}
```

soliton- commented:
IMO it's not helpful to call `aliasof` and `def_alias`/`mvt_alias` new names.

babaissarkar commented:
> IMO it's not helpful to call `aliasof` and `def_alias`/`mvt_alias` new names.
> 
> Well, `union_type` is not even `aliasof` but a bad mix that should not be exposed at all.

Can you link to docs for `aliasof` and `def_alias`/`mvt_alias`?

soliton- commented:
https://wiki.wesnoth.org/TerrainWML

babaissarkar commented:
> https://wiki.wesnoth.org/TerrainWML

Even after looking at `aliasof` and others, I don't understand how exposing `union_type` and others can be bad. `aliasof` does not look like it satisfies the requirements of #9969, which simply wanted to know the base terrain type of a given terrain (exact info as shown in the Help Browser).

babaissarkar commented:
@Discontinuum does this PR satisfy what you asked for?

babaissarkar commented:
> > > https://wiki.wesnoth.org/TerrainWML
> > 
> > 
> > Even after looking at `aliasof` and others, I don't understand how exposing `union_type` and others can be bad. `aliasof` does not look like it satisfies the requirements of #9969, which simply wanted to know the base terrain type of a given terrain (exactly same info as shown in the Help Browser), which is what I did: I exposed the same functions that the Help Browser uses to show the required information.
> 
> And the help is hopefully the only place that uses `union_type` since it's a mix of alias terrain that has very little meaning. Since it mixes in `vision_alias` it's rather misleading than anything else. I would honestly not use it in help either.
> 
> The actually used values are `mvt_alias` and `def_alias`. `aliasof` is a shortcut to set both to the same value. For retrieving the current values it's not really relevant. If there is a common need for the union of `mvt_alias` and `def_alias` that could be done but IMO it'd be more useful to add set operations to Lua if it's lacking there.

@soliton- I took a look at the actual code. Would you agree to this PR if I just expose `mvt_alias` and `def_alias` and drop `union_type`?

babaissarkar commented:
> > @soliton- I took a look at the actual code. Would you agree to this PR if I just expose `mvt_alias` and `def_alias` and drop `union_type`? I agree that union-ing the two lists would be doable in Lua if needed.
> 
> Yes, besides `union_type` this looks good. But expose them with those names. base and alias are different things here.

Updated. Dropped `union_type()` binding and renamed the others to `mvt_alias` and `def_alias`.

Discontinuum commented:
> does this PR satisfy what you asked for?

As far as I understand, it will show both forest and mountains if I ask for combined terrain like ``wesnoth.terrain_types["Hh^Fds"]``. If it's so, that's super cool.

A bit pity "best/worst of" proved to be too tricky, but just exposing the basic types is also a great start, thank you!

soliton- commented:
Note that `Hh^Fds` contains Hills and Forest as defense and movement alias not simply because it's a combined terrain of Hills and Forest but because Forest contains `_bas` as part of its aliases. Otherwise it would just be the overlay.

Also note that this currently returns IDs. Since `wesnoth.terrain_type` is indexed by terrain string instead it would probably make more sense to return that.

ProditorMagnus commented:
> Also note that this currently returns IDs. Since `wesnoth.terrain_type` is indexed by terrain string instead it would probably make more sense to return that.

Addon does not want to see _bas, addon needs to see "forest".

soliton- commented:
While `_bas` is indeed a valid terrain string currently I would strongly advise not to use it. Mainline does not so you will not see `_bas` there.

I would think you want to see `Ft` but you can certainly iterate all terrain types to find the one with id forest also.

soliton- commented:
No idea why you want that since there is barely a difference but you can trivially convert Ft to forest while the other way around is quite costly.

ProditorMagnus commented:
Forest is human readable, so that is the preferred form in ability definition.

soliton- commented:
You can get Forest via `wesnoth.terrain_types["Ft"].name`. A lot more costly to get to that from forest.

ProditorMagnus commented:
We dont need to know that internally forest is called Ft. Purpose of these issues is to add alternative to current requirement of hardcoding terrain codes.

soliton- commented:
And the purpose is fulfilled whether you use id or string or name or whatever? I fail to see what your point is. If it bothers you to look at a terrain string then just don't? Right now the interface is based on terrain strings though. That is my entire point. If you must work with terrain id then perhaps ask for an interface to use it.

ProditorMagnus commented:
We got the interface for terrain types in #9979.

```
$ wesnoth.terrain_types['Sm']
{id='quagmire',def_terrains={'swamp_water'},editor_image='terrain/swamp/mud-tile.png',light=0,icon='',mvt_terrains={'swamp_water'},keep=false,healing=0,basic_terrains={'swamp_water'},editor_name=Muddy Quagmire,village=false,castle=false,description=Swamp,name=Swamp}
```
Gives access to swamp_water. Your comment gave me impression you wanted to remove that from interface.

ProditorMagnus commented:
#9979 is intended to solve #9969, which requests the types that are used in movetypes. Adding some other random information to wesnoth.terrain_types is nice, but then this PR should not be marked as solution for #9969.

soliton- commented:
> > Also note that this currently returns IDs. Since wesnoth.terrain_type is indexed by terrain string instead it would probably make more sense to return that.
> 
> I think the intent is to use these to index a movetype rather than looking up the archetype terrain from `wesnoth.terrain_types`. While putting the terrain code instead of the ID here would be fine, I don't think it's necessary.

We can certainly make any use but terrain id difficult. No idea why we would but we can certainly do that.

> There was a mention of `vision_alias` earlier. If we're exposing these, we might as well expose that too, right? Is there a `jamming_alias`?

There is no `jamming_alias` and `vision_alias` "does nothing". The latter means that you get confusing help entries and editor display.

babaissarkar commented:
@soliton- @ProditorMagnus
Is it acceptable if I expose both ids and terrain strings? (adds two more tables) I cannot merge this until we get reach a consensus.

@ProditorMagnus 
Also, I fail to understand what the problem with terrain strings is. Given a terrain string, you can easily get the id:
```lua
local alias = wesnoth.terrain_types['Hh^Fds'].mvt_alias
if wesnoth.terrain_types[alias[0]].id == 'forest' then
...
```
You're still hardcoding `forest` instead of the corresponding terrain string. And you get extra terrain data as a bonus. The reverse is more costly, you have to loop over the entire `terrain_types` table to get extra terrain data if you even need any.

ProditorMagnus commented:
Looks good.

babaissarkar commented:
Demo code:
```lua
$ wesnoth.terrain_types['Sm']
{healing=0,description=Swamp,editor_image='terrain/swamp/mud-tile.png',def_alias={'St'},mvt_alias={'St'},keep=false,name=Swamp,village=false,castle=false,light=0,icon='',editor_name=Muddy Quagmire,id='quagmire'}
$ wesnoth.terrain_types['Sm'].mvt_alias[1]
'St'
$ wesnoth.terrain_types[wesnoth.terrain_types['Sm'].mvt_alias[1]]
{healing=0,description=Swamp,editor_image='',def_alias={'St'},mvt_alias={'St'},keep=false,name=Swamp,village=false,castle=false,light=0,icon='symbols/terrain_type_swamp_water',editor_name=Swamp,id='swamp_water'}
$ wesnoth.terrain_types[wesnoth.terrain_types['Sm'].mvt_alias[1]].id
'swamp_water'
$ wesnoth.terrain_types[wesnoth.terrain_types['Hh'].mvt_alias[1]].id
'hills'
$ wesnoth.terrain_types['Hh']
{healing=0,description=Hills,editor_image='terrain/hills/regular.png',def_alias={'Ht'},mvt_alias={'Ht'},keep=false,name=Hills,village=false,castle=false,light=0,icon='',editor_name=Regular Hills,id='hills_regular'}
$ wesnoth.terrain_types['Ht']
{healing=0,description=Hills,editor_image='',def_alias={'Ht'},mvt_alias={'Ht'},keep=false,name=Hills,village=false,castle=false,light=0,icon='symbols/terrain_type_hills',editor_name=Hills,id='hills'}
```
